packages/i18n/src/index.ts

```diff
@@ -157,10 +158,135 @@ export const createI18nContext = (
      */
     dict: (lang: string) => deepReadObject(dict, lang)
   };
-  return [translate, actions];
+  return [translate, actions as any];
 };
 
 type I18nContextInterface = ReturnType<typeof createI18nContext>;
 
 export const I18nContext = createContext<I18nContextInterface>({} as I18nContextInterface);
 export const useI18n = () => useContext(I18nContext);
+
+// -------------------------- Chained I18n -------------------
+export type I18nFormatOptions = Record<string, string | number>;
+
+interface I18nObject {
+  readonly [x: string]: string | ((...args: any) => any) | I18nObject;
+}
+
+type I18nFormatArgs = Record<string, string | number>;
+
+export type I18nPath<T extends I18nObject> = {
+  [K in keyof T]: T[K] extends I18nObject
+    ? I18nPath<T[K]>
+    : T[K] extends (options: infer OptionsArgs) => string
+    ? (options: OptionsArgs) => string
+    : (options?: I18nFormatArgs) => string;
+};
+
+export type Dictionaries<T extends I18nObject> = {
+  [key: string]: T;
+};
+
+type AddFunctionLengths<T> = T extends string
+  ? string
+  : T extends (...args: infer A) => unknown
+  ? T & { __length: A["length"] }
+  : {
+      [K in keyof T]: AddFunctionLengths<T[K]>;
+    };
+
+type RemoveFunctionLengths<T> = T extends string
+  ? string
+  : T extends (args: infer A) => infer R
+  ? (args: A) => R
+  : {
+      [K in keyof T]: RemoveFunctionLengths<T[K]>;
+    };
+
+type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void
+  ? I
+  : never;
+
+type GuaranteeIdenticalSignatures<T extends Dictionaries<I18nObject>> = Record<
+  keyof T,
+  RemoveFunctionLengths<UnionToIntersection<AddFunctionLengths<T[keyof T]>>>
+>;
+
+const buildI18nChain = <T extends I18nObject>(obj: T): I18nPath<T> => {
+  const keys = Object.keys(obj) as (keyof T)[];
+  const paths = keys.reduce((acc, key) => {
+    const value = obj[key] as any;
+    if (typeof value === "object") {
+      return {
+        ...acc,
+        [key]: buildI18nChain(value)
+      } as any;
+    } else if (typeof value === "function") {
+      return {
+        ...acc,
+        [key]: obj[key]
+      };
+    }
+    if (typeof value === "string") {
+      return {
+        ...acc,
+        [key]: (args: I18nFormatOptions) => {
+          return value.replace(/{{(.*?)}}/g, (_, key) => deepReadObject(args, key, ""));
+        }
+      };
+    } else {
+      throw new TypeError(
+        `Unsupported data format on the keys. Values must resolve to a string or a function that returns a string. Key name: "${
+          key as string
+        }"`
```

Comment:
This could be behind a DEV check
```suggestion
        process.env.DEV ? `Unsupported data format on the keys. Values must resolve to a string or a function that returns a string. Key name: "${
          key as string
        }"` : ""
```

Reply:
I can move it behind a DEV flag. This is a dead end that will cause it to crash if you tried to call it and would be typed as never.

Reply:
not saying to remove the error. But the message could be an empty string in production and a helpful message for the developer during development. No point shipping it to the end-user.